File: public/js/core/economics/reward.js
// reward.js — the accrual reward function, adopting the real Proof-of-
// Will formula the user's own prior work uses (YourMine's mine.js,
// calcClaimable()), not this project's earlier, simpler power-law
// r = K·b^α·q^β. Standard form:
//
//   r(b, q, qTotal, T) = (b · q^α) / [ln(qTotal^(β(1−T)) + C)]^γ
//
// Two adaptations from the original are deliberate, not incidental,
// both because AIWA domains never share a global clock while the
// original — a single Solana program every participant reads slots
// from — assumes one:
//
//   - qTotal ("protocol age" in the original: slots since a single
//     fixed global reference block, identical for every participant)
//     is this DOMAIN'S OWN total cadence epoch count here, not a
//     quantity shared across domains. Requiring every AIWA domain to
//     agree on one global reference epoch would reintroduce the exact
//     cross-domain synchronization dependency this architecture rules
//     out. Using each domain's own age preserves the formula's intent
//     — reward per action shrinks as a domain matures — without any
//     domain needing to know anything about any other domain's clock.
//   - The original enforces a minimum wait of 30 Solana slots
//     (~12 seconds) before any reward is claimable. AIWA's epoch is a
//     coarser, deployment-defined unit, not a fixed ~400ms slot, so the
//     literal number 30 would carry no equivalent meaning here.
//     Replaced with `minQ`, a deployment-chosen minimum epoch count,
//     defaulting to 1 (some non-zero cadence time must actually pass).
//
// b, q, and qTotal must be finite and >= 0 by construction; q = 0 still
// yields 0 whenever alpha > 0 for b > 0, the same "no instant reward"
// property the original power-law form had — cadence-derived economic
// time still governs, only the shape of the curve changed.

export class RewardError extends Error {}

/**
 * @typedef {{ alpha: number, beta: number, gamma: number, C: number, minQ: number }} RewardParams
 */

/**
 * Pure reward function, standard form above.
 *
 * @param {number} b - committed capital backing this claim (the original's S)
 * @param {number} q - epochs elapsed since this claim's q0 (Definition 10.1; the original's t)
 * @param {number} qTotal - this domain's own total cadence epoch count (the original's A, made domain-local — see file header)
 * @param {number} patienceRate - T, caller-chosen, clamped to [0, 0.4] exactly as the original does (Math.min(taxRate,40)/100)
 * @param {RewardParams} params
 * @returns {number}
 */
export function reward(b, q, qTotal, patienceRate, { alpha, beta, gamma, C, minQ }) {
  if (!Number.isFinite(b) || b < 0) throw new RewardError(`b must be a finite number >= 0, got ${b}`);
  if (!Number.isFinite(q) || q < 0) throw new RewardError(`q must be a finite number >= 0, got ${q}`);
  if (!Number.isFinite(qTotal) || qTotal < 0) throw new RewardError(`qTotal must be a finite number >= 0, got ${qTotal}`);
  if (![alpha, beta, gamma, C, minQ].every(Number.isFinite)) {
    throw new RewardError('alpha, beta, gamma, C, and minQ must all be finite numbers');
  }

  if (q < minQ) return 0;

  const T = Math.min(Math.max(patienceRate, 0), 0.4);
  const effQ = Math.max(1, q);
  const effQTotal = Math.max(1, qTotal);

  const numerator = effQ ** alpha * b;
  const inner = effQTotal ** (beta * (1 - T)) + C;
  if (inner <= 1) return 0;

  const denominator = Math.log(inner) ** gamma;
  if (!(denominator > 0) || !Number.isFinite(denominator) || !Number.isFinite(numerator)) return 0;

  const r = numerator / denominator;
  if (r < 0 || !Number.isFinite(r) || r > 1e12) return 0;
  return r;
}

/**
 * Derives q (elapsed economic epochs, Definition 10.1) for an accrual
 * event from the domain's current cadence state and the event's
 * acceptance epoch q_0. Unchanged from the prior revision.
 *
 * @param {import('./cadence.js').CadenceState} cadenceState
 * @param {string} domain
 * @param {number} q0
 * @returns {number}
 */
export function elapsedEpochs(cadenceState, domain, q0) {
  const currentEpoch = cadenceState.domains[domain]?.epoch ?? 0;
  return Math.max(0, currentEpoch - q0);
}

/**
 * qTotal — this domain's own total cadence epoch count right now, as
 * folded so far. See file header: this stands in for the original's
 * global "protocol age," deliberately made domain-local.
 *
 * @param {import('./cadence.js').CadenceState} cadenceState
 * @param {string} domain
 * @returns {number}
 */
export function domainAge(cadenceState, domain) {
  return cadenceState.domains[domain]?.epoch ?? 0;
}

File: rust-core/src/economics/reward.rs
use crate::economics::cadence::CadenceState;

/// Deployment-chosen constants for the reward function — mirror of
/// RewardParams in reward.js. See that file's header for the full
/// rationale: this is the real Proof-of-Will formula the project's
/// reference implementation uses, not the earlier power-law form.
#[derive(Debug, Clone, Copy)]
pub struct RewardParams {
    pub alpha: f64,
    pub beta: f64,
    pub gamma: f64,
    pub c: f64,
    pub min_q: f64,
}

#[derive(Debug, PartialEq)]
pub struct RewardError(pub String);

/// Pure reward function, standard form:
///
///   r(b, q, q_total, T) = (b · q^α) / [ln(q_total^(β(1−T)) + C)]^γ
///
/// Mirror of reward() in public/js/core/economics/reward.js — see that
/// file's header for the two deliberate adaptations from the original
/// formula (q_total is domain-local, not a shared global "protocol
/// age"; min_q replaces a Solana-slot-specific minimum wait with a
/// deployment-chosen epoch count).
pub fn reward(b: f64, q: f64, q_total: f64, patience_rate: f64, params: RewardParams) -> Result<f64, RewardError> {
    if !b.is_finite() || b < 0.0 {
        return Err(RewardError(format!("b must be a finite number >= 0, got {b}")));
    }
    if !q.is_finite() || q < 0.0 {
        return Err(RewardError(format!("q must be a finite number >= 0, got {q}")));
    }
    if !q_total.is_finite() || q_total < 0.0 {
        return Err(RewardError(format!("q_total must be a finite number >= 0, got {q_total}")));
    }
    if ![params.alpha, params.beta, params.gamma, params.c, params.min_q].iter().all(|v| v.is_finite()) {
        return Err(RewardError("alpha, beta, gamma, C, and min_q must all be finite numbers".to_string()));
    }

    if q < params.min_q {
        return Ok(0.0);
    }

    let t = patience_rate.max(0.0).min(0.4);
    let eff_q = q.max(1.0);
    let eff_q_total = q_total.max(1.0);

    let numerator = eff_q.powf(params.alpha) * b;
    let inner = eff_q_total.powf(params.beta * (1.0 - t)) + params.c;
    if inner <= 1.0 {
        return Ok(0.0);
    }

    let denominator = inner.ln().powf(params.gamma);
    if !(denominator > 0.0) || !denominator.is_finite() || !numerator.is_finite() {
        return Ok(0.0);
    }

    let r = numerator / denominator;
    if r < 0.0 || !r.is_finite() || r > 1e12 {
        return Ok(0.0);
    }
    Ok(r)
}

/// Derives q (elapsed economic epochs, Definition 10.1) for an accrual
/// event from the domain's current cadence state and the event's
/// acceptance epoch q_0. Unchanged from the prior revision.
pub fn elapsed_epochs(cadence_state: &CadenceState, domain: &str, q0: u64) -> u64 {
    let current_epoch = cadence_state.domains.get(domain).map(|d| d.epoch).unwrap_or(0);
    current_epoch.saturating_sub(q0)
}

/// q_total — this domain's own total cadence epoch count right now, as
/// folded so far. Mirror of domainAge() in reward.js: stands in for
/// the original formula's global "protocol age," deliberately made
/// domain-local.
pub fn domain_age(cadence_state: &CadenceState, domain: &str) -> u64 {
    cadence_state.domains.get(domain).map(|d| d.epoch).unwrap_or(0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::economics::cadence::DomainCadenceState;

    // The real reference implementation's actual constants
    // (mine.js's calcClaimable(): alpha=1.1, beta=2.2, gamma=3, C=33^3).
    fn real_params() -> RewardParams {
        RewardParams { alpha: 1.1, beta: 2.2, gamma: 3.0, c: 33f64.powi(3), min_q: 1.0 }
    }

    #[test]
    fn matches_hand_computed_value_using_real_formula_constants() {
        // b=1, q=100, q_total=100, T=0 — see reward.test.mjs for the
        // independently-verified expected value.
        let r = reward(1.0, 100.0, 100.0, 0.0, real_params()).unwrap();
        assert!((r - 0.11844290947765648).abs() < 1e-9);
    }

    #[test]
    fn larger_b_scales_numerator_linearly() {
        let r1 = reward(1.0, 100.0, 100.0, 0.0, real_params()).unwrap();
        let r10 = reward(10.0, 100.0, 100.0, 0.0, real_params()).unwrap();
        assert!((r10 - r1 * 10.0).abs() < 1e-6);
    }

    #[test]
    fn q_below_min_q_yields_zero_reward() {
        let r = reward(10.0, 0.0, 100.0, 0.0, real_params()).unwrap();
        assert_eq!(r, 0.0);
    }

    #[test]
    fn q_at_or_above_min_q_yields_positive_reward() {
        let r = reward(10.0, 1.0, 1.0, 0.0, real_params()).unwrap();
        assert!(r > 0.0);
    }

    #[test]
    fn higher_patience_rate_changes_the_reward() {
        let r_no_patience = reward(10.0, 100.0, 100.0, 0.0, real_params()).unwrap();
        let r_max_patience = reward(10.0, 100.0, 100.0, 0.4, real_params()).unwrap();
        assert_ne!(r_no_patience, r_max_patience);
    }

    #[test]
    fn patience_rate_clamped_to_0_4() {
        let r_clamped = reward(10.0, 100.0, 100.0, 0.4, real_params()).unwrap();
        let r_overclaimed = reward(10.0, 100.0, 100.0, 0.9, real_params()).unwrap();
        assert_eq!(r_clamped, r_overclaimed);
    }

    #[test]
    fn negative_patience_rate_clamped_to_0() {
        let r_zero = reward(10.0, 100.0, 100.0, 0.0, real_params()).unwrap();
        let r_negative = reward(10.0, 100.0, 100.0, -5.0, real_params()).unwrap();
        assert_eq!(r_zero, r_negative);
    }

    #[test]
    fn negative_b_is_rejected() {
        assert!(reward(-1.0, 1.0, 1.0, 0.0, real_params()).is_err());
    }

    #[test]
    fn negative_q_is_rejected() {
        assert!(reward(1.0, -1.0, 1.0, 0.0, real_params()).is_err());
    }

    #[test]
    fn negative_q_total_is_rejected() {
        assert!(reward(1.0, 1.0, -1.0, 0.0, real_params()).is_err());
    }

    #[test]
    fn non_finite_params_are_rejected() {
        let mut p = real_params();
        p.alpha = f64::NAN;
        assert!(reward(1.0, 1.0, 1.0, 0.0, p).is_err());
        let mut p2 = real_params();
        p2.gamma = f64::INFINITY;
        assert!(reward(1.0, 1.0, 1.0, 0.0, p2).is_err());
    }

    #[test]
    fn reward_never_negative_or_non_finite_for_valid_inputs() {
        let r = reward(1e6, 5000.0, 5000.0, 0.2, real_params()).unwrap();
        assert!(r.is_finite() && r >= 0.0);
    }

    #[test]
    fn elapsed_epochs_reads_current_domain_epoch() {
        let mut state = CadenceState::new();
        state.domains.insert("d1".to_string(), DomainCadenceState { epoch: 7, last_id: Some("c7".to_string()), vdf_output: None });
        assert_eq!(elapsed_epochs(&state, "d1", 3), 4);
    }

    #[test]
    fn elapsed_epochs_never_goes_negative() {
        let mut state = CadenceState::new();
        state.domains.insert("d1".to_string(), DomainCadenceState { epoch: 2, last_id: Some("c2".to_string()), vdf_output: None });
        assert_eq!(elapsed_epochs(&state, "d1", 5), 0);
    }

    #[test]
    fn elapsed_epochs_for_unknown_domain_is_zero() {
        let state = CadenceState::new();
        assert_eq!(elapsed_epochs(&state, "unknown", 0), 0);
    }

    #[test]
    fn domain_age_reads_current_epoch() {
        let mut state = CadenceState::new();
        state.domains.insert("d1".to_string(), DomainCadenceState { epoch: 42, last_id: Some("c42".to_string()), vdf_output: None });
        assert_eq!(domain_age(&state, "d1"), 42);
    }

    #[test]
    fn domain_age_for_unknown_domain_is_zero() {
        let state = CadenceState::new();
        assert_eq!(domain_age(&state, "unknown"), 0);
    }
}
